src/pybind/mgr/dashboard_v2/HACKING.rst

```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
 that setting from the CLI and vice-versa.
 
+
+How to run a controller read-write operation asynchronously?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some controllers might need to execute operations that alter the state of the
+Ceph cluster. These operations might take some time to execute and to maintain
+a good user experience in the Web UI, we need to run those operations
+asynchronously and return immediatly to frontend some information that the
+operations are running in the background.
+
+To help in the development of the above scenario we added the support for
+asynchronous tasks. To trigger the execution of an asynchronous task we must
+use the follwoing class method of the ``TaskManager`` class::
+
+  import ..tools import TaskManager
+  # ...
+  TaskManager.run(namespace, metadata, func, args, kwargs)
+
+* ``namespace`` is a string that can be used to group tasks. For instance
+  for RBD image creation tasks we could specify ``"rbd/create"`` as the
+  namespace, or conversly ``"rbd/remove"`` for RBD image removal tasks.
+
+* ``metadata`` is a dictionary where we can store key-value pairs that
+  characterize the task. For instance, when creating a task for creating
+  RBD images we can specify the metadata argument as
+  ``{'pool_name': "rbd", image_name': "test-img"}``.
+
+* ``func`` is the python function that implements the operation code, which
+  will be executed asynchronously.
+
+* ``args`` and ``kwargs`` are the positional and named argguments that will be
+  passed to ``func`` when the task manager starts its execution.
+
+The ``TaskManager.run`` method triggers the asynchronous execution of function
+``func`` and returns an ``AsyncTask`` object.
+The ``AsyncTask`` provides the public method ``AsyncTask.wait(timeout)``, which
+can be used to wait for the task to complete up to a timeout defined in seconds
+and provided as an argument. If not argument is provided the ``wait`` method
+blocks until the task is finished.
+
+The ``AsyncTask.wait`` is very useful for tasks that usually are fast
+to execute but that sometimes may take a long time to run.
+The return value of the ``AsyncTask.wait`` method is a pair ``(state, value)``
+where ``state`` is an integer with following possible values:
+
+* ``VALUE_DONE = 0``
+* ``VALUE_EXECUTING = 1``
+* ``VALUE_EXCEPTION = 2``
+
+The ``value`` will store the result of the execution of function ``func`` if
+``state == VALUE_DONE``. If ``state == VALUE_EXECUTING`` then
+``value == None``, and if ``state == VALUE_EXCEPTION`` then ``value`` stores
+the exception object raised by the execution of function ``func``.
+
+The pair ``(namespace, metadata)`` should univocally identify the task being
+run, which means that if you try to trigger a new task that matches the same
+``(namespace, metadata)`` pair of the currently running task, then the new task
+is not created and you get the task object of the current running task.
```

Comment:
This is bad. Consider:
```
def set_pg_count_per_pool(pool, count):
    CephService.send_message(....)
   ...

TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))
TaskManager.run('set_pg_count_per_pool', 12, lambda: set_pg_count_per_pool(mypool, 12))
TaskManager.run('set_pg_count_per_pool', 42, lambda: set_pg_count_per_pool(mypool, 42))
```

In the end, the pool would have 12 pgs, instead of 42. But the user would have a successful answer from the UI. 

Instead, `TaskManager` should *per default* create every task independent of any parameters.

Reply:
First, the developer needs to follow the contract of `TaskManager.run` as stated in `HACKING.rst`, which says that `(namespace, metadata)` must unequivocally identify a task, and as also stated in the documentation, he knows that if he triggers a new task with the same `(namespace, metadata)` as an already executing task that he will not be triggering a new task.
Therefore, according to the documentation the developer should not write the code you used as an example.

Moreover, the above example is a good example of a bad specification of the `(namespace, metadata)`. For the above example, the `(namespace, metadata)` could be: `("osd/pool/set_pg_count", {"pool_name": mypool})`. This way, the frontend could easily map the executing tasks to the "pool" table in the "Pools Page" (that does not exist yet).

Reply:
> Moreover, the above example is a good example of a bad specification of the `(namespace, metadata)`. For the above example, the `(namespace, metadata)` could be: `("osd/pool/set_pg_count", {"pool_name": mypool})`. This way, the frontend could easily map the executing tasks to the "pool" table in the "Pools Page" (that does not exist yet).

Your example supports my arguments. `("osd/pool/set_pg_count", {"pool_name": mypool})` will make things even worse by silently ignoring _all_ `set_pg_count` tasks while the first one is still running. Which means that a developer has to be careful in designing a task's metadata, in order to not run into the trap of being silently ignored. This shouldn't be necessary and looks to me like a bug.

Finally, `"osd/pool/set_pg_count"` is exactly something that I would think of a perfect example of a task name: it consists of the controller class name and the method name.

Reply:
@sebastian-philipp you can't use nonsensical example as an argument. Why would the user issue three asynchronous "set_pg_count" operations targeting the same pool?

The `("osd/pool/set_pg_count", {"pool_name": mypool})` specification actually avoids someone from doing such mistake, by guaranteeing that only one "set_pg_count" operation is running at a time.

Reply:
I'm not entirely following what the divisiveness here is. Correct me if I'm wrong:

1. We have a queue of tasks being executed or pending execution, identified by a given key consisting of `(namespace, metadata)`.

2. A developer queuing one task `(foo/bar, 1337)` will be able to later check on said task using the previously defined key.

3. A developer queuing multiple tasks `(foo/bar, 1337)` and `(foo/baz, 1337)` will in fact be queuing two different tasks, which can be checked up using their individual keys.

4. A developer queuing multiple tasks `(foo/bar, 1337)` will in fact be queuing ONLY one task.

5. Tasks being queued with the same key `(foo/bar, 1337)` while an already existing task with said key is executing, will be considered idempotent requests and will end up waiting for the same, original task to finish.

6. A developer checking up on a task `(foo/bar, 1337)` multiple times will always get the same response from the queue, whether it's still executing or finished execution, until a given number of seconds/minutes has passed.

Am I missing something?

Even although I haven't read the code in its entirety, the architectural aspect of how this works makes a lot of sense to me. Surely a developer will be able to abuse it, but that's why critical code review should be enforced.

Reply:
I'd like to get @ricardoasmarques take on this, as he's currently working on the implementation that will have to consume this API. We should make sure that his requirements are being met and addressed.

Reply:
From the consumer/front-end point of view, it makes sense to have a unique identifier composed by `namespace` and `metadata` (we can call it `name` or something else if you prefer to have a different name, but the behaviour should remain the same).

If I understand correctly, to follow @sebastian-philipp suggestion, we will need to have a unique identifier generated by server (e.g. a unique numeric id), and we don't want to depend on that numeric id on the frontend because this will make it very hard to deal with cases where user do a full page refresh or move to a diferent browser during the execution of a task.


Taking into account the experience we had with openATTIC while implementing iSCSI and NFS service management, we simply ignored the numeric taskid on the frontend when displaying the "Starting..." and "Stopping..." messages to guarantee that the current status is always displayed even after a full page refresh.